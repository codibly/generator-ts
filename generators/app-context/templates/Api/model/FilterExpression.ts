/**
 * Basic operands
 */
export interface IdentifierOperand {
  identifier: string;
}
export interface LiteralOperand {
  literal: any;
}

/**
 * Basic expressions
 */
export interface IdentifierExpression {
  [0]: IdentifierOperand;
  length: 1;
}
export interface IdentifierLiteralExpression {
  [0]: IdentifierOperand;
  [1]: LiteralOperand;
  length: 2;
}

/**
 * Terminal expressions
 */
export interface NullExpression {
  null: IdentifierExpression;
}
export interface NotNullExpression {
  notNull: IdentifierExpression;
}
export interface LtExpression {
  lt: IdentifierLiteralExpression;
}
export interface LteExpression {
  lte: IdentifierLiteralExpression;
}
export interface GtExpression {
  gt: IdentifierLiteralExpression;
}
export interface GteExpression {
  gte: IdentifierLiteralExpression;
}
export interface InExpression {
  in: IdentifierLiteralExpression;
}
export interface EqExpression {
  eq: IdentifierLiteralExpression;
}
export interface LikeExpression {
  like: IdentifierLiteralExpression;
}
export type TerminalExpression =
  | NullExpression
  | NotNullExpression
  | LtExpression
  | LteExpression
  | GtExpression
  | GteExpression
  | InExpression
  | EqExpression
  | LikeExpression;

/**
 * Logic expressions
 */
export interface AndExpression {
  and: FilterExpression[];
}
export interface OrExpression {
  or: FilterExpression[];
}
export type LogicExpression = AndExpression | OrExpression;

/**
 * Filter expression
 */
export type FilterExpression = LogicExpression | TerminalExpression;

export namespace FilterExpression {
  // type guards
  export function isLikeExpression(filter: FilterExpression): filter is LikeExpression {
    return !!(filter as LikeExpression).like;
  }

  // builders
  export enum LikeMode {
    STARTS_WITH = 'STARTS_WITH',
    CONTAINS = 'CONTAINS',
    EQUALS = 'EQUALS',
    ENDS_WITH = 'ENDS_WITH'
  }

  export function likeLiteral(literal: string, mode: LikeMode): string {
    switch (mode) {
      case LikeMode.STARTS_WITH:
        return `${literal}%`;

      case LikeMode.ENDS_WITH:
        return `%${literal}`;

      case LikeMode.CONTAINS:
        return `%${literal}%`;

      case LikeMode.EQUALS:
      default:
        return literal;
    }
  }

  // we can't use "in" as it's keyword in the JavaScript
  export function oneOf(identifier: string, literal: any[]): InExpression {
    return {
      in: [{ identifier }, { literal }]
    };
  }

  export function eq(identifier: string, literal: any): EqExpression {
    return {
      eq: [{ identifier }, { literal }]
    };
  }

  export function like(
    identifier: string,
    query: string,
    mode: LikeMode = LikeMode.CONTAINS
  ): LikeExpression {
    return {
      like: [{ identifier }, { literal: likeLiteral(query, mode) }]
    };
  }

  export function or(filters: FilterExpression[]): FilterExpression {
    if (filters.length === 1) {
      return filters[0];
    } else {
      return {
        or: filters
      };
    }
  }

  export function and(filters: FilterExpression[]): FilterExpression {
    if (filters.length === 1) {
      return filters[0];
    } else {
      return {
        and: filters
      };
    }
  }
}
