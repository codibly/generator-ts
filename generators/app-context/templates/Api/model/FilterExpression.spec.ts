import { FilterExpression } from './FilterExpression';

describe('FilterExpression', () => {
  it.each([
    FilterExpression.LikeMode.STARTS_WITH,
    FilterExpression.LikeMode.ENDS_WITH,
    FilterExpression.LikeMode.CONTAINS,
    FilterExpression.LikeMode.EQUALS
  ])('should create like literal for mode %p', (mode) => {
    expect(FilterExpression.likeLiteral('test', mode)).toMatchSnapshot();
  });

  it('should create "in" expression', () => {
    expect(FilterExpression.oneOf('field', ['a', 'b', 'c'])).toMatchSnapshot();
  });

  it('should create "eq" expression', () => {
    expect(FilterExpression.eq('field', 'test')).toMatchSnapshot();
  });

  it.each([
    FilterExpression.LikeMode.STARTS_WITH,
    FilterExpression.LikeMode.ENDS_WITH,
    FilterExpression.LikeMode.CONTAINS,
    FilterExpression.LikeMode.EQUALS
  ])('should create "like" expression for %p mode', (mode) => {
    expect(FilterExpression.like('field', 'test', mode)).toMatchSnapshot();
  });

  it('should create "or" expression for single input', () => {
    expect(FilterExpression.or([FilterExpression.eq('field', 'test')])).toMatchSnapshot();
  });

  it('should create "or" expression for multi input', () => {
    expect(
      FilterExpression.or([
        FilterExpression.eq('field', 'test'),
        FilterExpression.eq('field', 'abc'),
        FilterExpression.like('field', 'efg')
      ])
    ).toMatchSnapshot();
  });

  it('should create "and" expression for single input', () => {
    expect(FilterExpression.and([FilterExpression.eq('field', 'test')])).toMatchSnapshot();
  });

  it('should create "and" expression for multi input', () => {
    expect(
      FilterExpression.and([
        FilterExpression.eq('field', 'test'),
        FilterExpression.eq('field', 'abc'),
        FilterExpression.like('field', 'efg')
      ])
    ).toMatchSnapshot();
  });
});
