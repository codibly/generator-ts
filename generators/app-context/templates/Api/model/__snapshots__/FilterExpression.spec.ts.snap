// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterExpression should create "and" expression for multi input 1`] = `
Object {
  "and": Array [
    Object {
      "eq": Array [
        Object {
          "identifier": "field",
        },
        Object {
          "literal": "test",
        },
      ],
    },
    Object {
      "eq": Array [
        Object {
          "identifier": "field",
        },
        Object {
          "literal": "abc",
        },
      ],
    },
    Object {
      "like": Array [
        Object {
          "identifier": "field",
        },
        Object {
          "literal": "%efg%",
        },
      ],
    },
  ],
}
`;

exports[`FilterExpression should create "and" expression for single input 1`] = `
Object {
  "eq": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "test",
    },
  ],
}
`;

exports[`FilterExpression should create "eq" expression 1`] = `
Object {
  "eq": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "test",
    },
  ],
}
`;

exports[`FilterExpression should create "in" expression 1`] = `
Object {
  "in": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": Array [
        "a",
        "b",
        "c",
      ],
    },
  ],
}
`;

exports[`FilterExpression should create "like" expression for "CONTAINS" mode 1`] = `
Object {
  "like": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "%test%",
    },
  ],
}
`;

exports[`FilterExpression should create "like" expression for "ENDS_WITH" mode 1`] = `
Object {
  "like": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "%test",
    },
  ],
}
`;

exports[`FilterExpression should create "like" expression for "EQUALS" mode 1`] = `
Object {
  "like": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "test",
    },
  ],
}
`;

exports[`FilterExpression should create "like" expression for "STARTS_WITH" mode 1`] = `
Object {
  "like": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "test%",
    },
  ],
}
`;

exports[`FilterExpression should create "or" expression for multi input 1`] = `
Object {
  "or": Array [
    Object {
      "eq": Array [
        Object {
          "identifier": "field",
        },
        Object {
          "literal": "test",
        },
      ],
    },
    Object {
      "eq": Array [
        Object {
          "identifier": "field",
        },
        Object {
          "literal": "abc",
        },
      ],
    },
    Object {
      "like": Array [
        Object {
          "identifier": "field",
        },
        Object {
          "literal": "%efg%",
        },
      ],
    },
  ],
}
`;

exports[`FilterExpression should create "or" expression for single input 1`] = `
Object {
  "eq": Array [
    Object {
      "identifier": "field",
    },
    Object {
      "literal": "test",
    },
  ],
}
`;

exports[`FilterExpression should create like literal for mode "CONTAINS" 1`] = `"%test%"`;

exports[`FilterExpression should create like literal for mode "ENDS_WITH" 1`] = `"%test"`;

exports[`FilterExpression should create like literal for mode "EQUALS" 1`] = `"test"`;

exports[`FilterExpression should create like literal for mode "STARTS_WITH" 1`] = `"test%"`;
